### PR TITLE
Exclude data outside of encoding_matrix

### LINKED ISSUE #217: Mismatch encoding limits - encoding matrix
The encoding matrix describes the size of the k-space which needs to be encoded based on the resolution and FOV settings. Commonly data acquisition is then carried out in this matrix (or smaller due to partial fourier or undersampling). This means the encoding limits are smaller or equal to the matrix. 

There are cases where this is not the case. If the encoding matrix along PE is 188 but the number of samples per shot (e.g. PE points acquired after a prep-pulse or a trigger signal) is 20, then actually 200 PE samples will be obtained leading to the encoding limits of [0, 199]. 

For Cartesian sampling this currently breaks then e.g. here https://github.com/PTB-MR/mrpro/blob/9abbd8c90563d5b922f42757314d2902de615627/src/mrpro/operators/_CartesianSamplingOp.py#L128.

Easy solution would be to ignore points outside of encoding_matrix but of course it would be nicer to extend the encoding_matrix to include these points. This would require image-based interpolation to achieve to correct resolution.

### PR DESCRIPTION
**Pull request by ckolbPTB**
closes #217

### CODE REVIEW

**Comment by ckolbPTB** [comment]
any ideas of what to do now with the exclude info are very welcome

File: src/mrpro/operators/_CartesianSamplingOp.py [hunk 1]
```diff
@@ -78,9 +80,37 @@ def __init__(self, encoding_matrix: SpatialDimension[int], traj: KTrajectory) ->
...
+            & (kz_idx < sorted_grid_shape.z)
+        )
+        if torch.any(~inside_encoding_matrix):
+            warnings.warn(
+                'K-space points lie outside of the encoding_matrix and will be ignored.'
+                ' Increase the encoding_matrix to include these points.',
+                stacklevel=2,
+            )
+
+            inside_encoding_matrix = rearrange(inside_encoding_matrix, '... kz ky kx -> ... 1 (kz ky kx)')
```
Comment (ckolbPTB) [inline src/mrpro/operators/_CartesianSamplingOp.py]: This part is not very nice. Maybe somebody has a better idea of how to do create the indices.

Comment (fzimmermann89) [reply src/mrpro/operators/_CartesianSamplingOp.py]: i have a suggestion using non-zero

**Comment by fzimmermann89** [comment]
- rebased to main.
- made test more difficult: two oversampled axes and other dimension (which failed)
- change the index creation to use non-zero
- fix for other dimension, unbreaks more difficult test. uses unsqueeze_left
- tiny performance thing: order of adding broadcasted tensors to figure out the inside_mask
- removed the bool marking telling if inside_matrix indices exist, and instead make it the indices optional/None.
- changed some comments (minor)

@ckolbPTB  did I mess anything up?

I also think there might be a nicer way to do this, but this works for now, lets see if it ever actually becomes a performance bottleneck.

**Review by fzimmermann89 — APPROVED** [review]

**Comment by ckolbPTB** [comment]
verified on real data

**Comment by fzimmermann89** [comment]
Thank you for checking on real data! I will merge?

**Pull request merged**